File: bank-account/src/main/java/org/suggs/sandbox/eventsourcing/bankaccount/accounts/commandhandler/CreateAccountCommandHandler.java
package org.suggs.sandbox.eventsourcing.bankaccount.accounts.commandhandler;

import com.google.common.eventbus.Subscribe;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.suggs.sandbox.eventsourcing.bankaccount.accounts.bus.CommandBus;
import org.suggs.sandbox.eventsourcing.bankaccount.accounts.domain.commands.ConfirmNewAccountToClientCommand;
import org.suggs.sandbox.eventsourcing.bankaccount.accounts.domain.commands.CreateAccountCommand;
import org.suggs.sandbox.eventsourcing.bankaccount.accounts.domain.events.AccountCreatedEvent;
import org.suggs.sandbox.eventsourcing.bankaccount.accounts.repository.EventRepository;

import java.math.BigDecimal;
import java.util.UUID;

public class CreateAccountCommandHandler implements CommandHandler<CreateAccountCommand> {

    private Logger LOG = LoggerFactory.getLogger(CreateAccountCommandHandler.class);

    private final CommandBus commandBus;
    private final EventRepository eventRepository;

    public CreateAccountCommandHandler(CommandBus commandBus, EventRepository eventRepository) {
        this.commandBus = commandBus;
        this.eventRepository = eventRepository;
    }

    @Override
    @Subscribe
    public void handle(CreateAccountCommand aCommand) {
        UUID newAccountNumber = createNewAccount(aCommand.getCreateAccountRequestId(), aCommand.getInitialBalance());
        commandBus.publish(new ConfirmNewAccountToClientCommand(aCommand.getCreateAccountRequestId(), newAccountNumber));
        eventRepository.save(new AccountCreatedEvent(aCommand.getCreateAccountRequestId(), newAccountNumber, aCommand.getInitialBalance()));
    }

    private UUID createNewAccount(UUID requestId, BigDecimal anInitialBalance) {
        LOG.debug("########### Creating a new account number for the request " + requestId.toString() + " with an initial balance of " + anInitialBalance);
        return UUID.randomUUID();
    }
}

File: bank-account/src/main/kotlin/org/suggs/sandbox/eventsourcing/bankaccount/accounts/commandhandler/CreateAccountCommandHandler.kt
package org.suggs.sandbox.eventsourcing.bankaccount.accounts.commandhandler

import com.google.common.eventbus.AllowConcurrentEvents
import com.google.common.eventbus.Subscribe
import org.slf4j.LoggerFactory
import org.suggs.sandbox.eventsourcing.bankaccount.accounts.bus.CommandBus
import org.suggs.sandbox.eventsourcing.bankaccount.accounts.domain.command.ConfirmAccountToClient
import org.suggs.sandbox.eventsourcing.bankaccount.accounts.domain.command.CreateAccount
import org.suggs.sandbox.eventsourcing.bankaccount.accounts.domain.event.AccountCreated
import org.suggs.sandbox.eventsourcing.bankaccount.accounts.repository.EventRepository
import java.math.BigDecimal
import java.util.*

class CreateAccountCommandHandler(private val eventRepo: EventRepository, private val commandBus: CommandBus) : CommandHandler<CreateAccount> {

    companion object {
        private val log = LoggerFactory.getLogger(this::class.java)
    }

    @Subscribe
    @AllowConcurrentEvents
    override fun handle(command: CreateAccount) {
        val accountNumber = createNewAccount(command.requestId, command.initialBalance)
        eventRepo.save(AccountCreated(command.requestId, accountNumber, command.initialBalance))
        commandBus.publish(ConfirmAccountToClient(command.requestId, accountNumber))
    }

    private fun createNewAccount(customerId: UUID, openingBalance: BigDecimal): UUID {
        log.debug("Creating a new account for the client [$customerId] with an opening balance of [$openingBalance]")
        return UUID.randomUUID()
    }
}